Replace browserify-rails with rails-webpacker

jmhooper opened the pull request:
**Why**: rails-webpacker compiles the assets much faster than
browserify-rails. With browserify-rails, our assets take so long to
compile that the initial requests to the idp time out locally. I have
not observed that with webpacker.

Additionally, rails-webpacker is recommended by the maintainer of
browserify-rails for projects that have already made the step towards
using commonjs modules.

One tradeoff of this is that rails-webpacker expects yarn, so we'll need
to move from npm to yarn for this to work.

Another tradeoff is that teaspoon does not appear to play nicely with
rails-webpacker, so I removed it and replaced it with a bare-metal
mocha.

Hi! Before submitting your PR for review, and/or before merging it, please
go through the following checklist:

- [x] For DB changes, check for missing indexes, check to see if the changes
affect other apps (such as the dashboard), make sure the DB columns in the
various environments are properly populated, coordinate with devops, plan
migrations in separate steps.

- [x] For route changes, make sure GET requests don't change state or result in
destructive behavior. GET requests should only result in information being
read, not written.

- [x] For encryption changes, make sure it is compatible with data that was
encrypted with the old code.

- [x] Do not disable Rubocop or Reek offenses unless you are absolutely sure
they are false positives. If you're not sure how to fix the offense, please
ask a teammate.

- [x] When reading data, write tests for nil values, empty strings,
and invalid formats.

- [x] When calling `redirect_to` in a controller, use `_url`, not `_path`.

- [x] When adding user data to the session, use the `user_session` helper
instead of the `session` helper so the data does not persist beyond the user's
session.

jmhooper commented on config/deploy.rb:
```diff
@@ -49,11 +49,11 @@
     on roles(:app, :web), in: :sequence do
       within release_path do
-        execute :npm, 'install'
+        execute :yarn, 'install'
```
Before merging, we'll want to work with devops to make sure yarn is available.

tbaxter-18f commented:
My general preference would be to run webpack directly as needed, but I think these are really solid changes.

jmhooper commented:
@tbaxter-18f: The downside of running webpack directly is we loose a few of the benefits of the assets pipeline, like the use of tag helpers to load thumbprinted assets. I've managed to work around that by bundling assets into the asset load path, but that gets messy and with larger bundles can mean we see more performance issues.

jmhooper commented:
Also, note that the code climate issues are related to files I moved, not files I've changed. I can fix the issues in them if we think it is a priority, but I'd prefer to keep the diff here small.

jmhooper commented:
I'm going to move this down to WIP. Looks like we'll need to move up to node v6 or v8 to use `@rails/webpacker` and yarn.

monfresh commented:
@jmhooper Wanna rebase with master?

monfresh commented on .circleci/config.yml:
```diff
@@ -56,14 +56,15 @@ jobs:
             bin/generate-example-keys
             bundle exec rake db:setup --trace
             bundle exec rake assets:precompile
+            bundle exec rake db:setup --trace
```
Looks like we already have this on line 65.

monfresh commented on app/views/layouts/application.html.slim:
```diff
@@ -21,9 +21,8 @@ html lang="#{I18n.locale}" class='no-js'
-    == javascript_include_tag 'misc/i18n-strings'
-    - unless controller_name == 'home'
-      == javascript_include_tag 'application'
+    == javascript_include_tag 'i18n-strings'
```
should this be `javascript_pack_tag`?

jmhooper replied:
No, the `i18n-strings` use ERB so I've left it in the app directory for now. One thing to think about after we handle this may be to move that into a template like we do with the session timeout javascript.

monfresh commented on app/views/layouts/application.html.slim:
```diff
@@ -21,9 +21,8 @@ html lang="#{I18n.locale}" class='no-js'
-    - unless controller_name == 'home'
-      == javascript_include_tag 'application'
+    == javascript_include_tag 'i18n-strings'
+    == javascript_pack_tag 'application'
```
We don't need the controller conditional anymore?

jmhooper replied:
I don't think a "home" controller exists anymore?

monfresh replied:
heh. you're right!

jmhooper commented:
@monfresh: Addressed the things you mentioned and brought this up to date with master. Should be good to go now.

Also note the synk issue is a false positive IIRC. It relates to using the dep in a Node.js app which we are not doing.
